Document Outline: Add check for the main element

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR description was updated 2025-04-12.

This PR updates the `DocumentOutline `component to include a check for blocks with the `main` tagName attribute.
In block themes, when a template is being edited or when the option "Show template" is enabled, 
the outline displays headings and blocks with the main element, including nested blocks.

If there is more than one main, there is a warning with this text below every listed main element:
"Your template should have exactly one main element. Adjust the HTML element in the Advanced panel of the block."

If the main is missing, a warning at the top of the document outline says:
"The main element is missing. Select the block that contains your most important content and add the main HTML element in the Advanced panel."

(The warning for incorrect heading levels remains the same.)

Closes https://github.com/WordPress/gutenberg/issues/35354

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Having a single `<main> ` element in a template is important for accessibility, but also for the Zoom Out feature.
Having more than one is not valid HTML.
The skip link feature is only enabled on templates that has a `<main>` element.
See the issue for more details.

## Testing Instructions
Activate a block theme
Go to Appearance > Editor and open any template.
Open the List View
Open the Document Outline tab.

Locate or add a block that has the `<main>`  HTML element: This is usually a group, but any block with the `tagName` attribute could be using it.
To add the main HTML element using the interface option, select a container block, open the Block Settings Sidebar, and open the Advanced panel. Select the main in the HTML element option.
Confirm that the main element is showing in the Document Outline.
Confirm that selecting the item in the Outline selects the block.

Duplicate the block that has the `<main>`  element and confirm if both main elements are listed in the Outline.

Move one of the mains inside the other. Add a few heading blocks. Confirm that both the nested main and the nested heading blocks show in the outline.

Delete all blocks with a `<main>` element and confirm if there is a text at the top of the Outline that explains that the template should have one main.

Test that there are no regressions in the post editor. 
When the option "show template" is off, the document outline should not list the main element, or warnings for the main element.
When the option "show template" is enabled, the document outline should list the main element and warnings for the main element.

Test that there are no regressions in the post editor when choosing the option "Edit template".
When the template is edited, the post editor should show warnings for the main element.

Activate a classic theme.
Test that there are no regressions in the post editor. The post editor should not show warnings for the main element.

## Screenshots or screencast <!-- if applicable -->

### Before
No warnings:
![The document outline without warnings, showing a few headings](https://github.com/user-attachments/assets/773e0145-a6fc-4fee-ba73-657067ed797b)

### After

New screenshot, including the notice at the top of the editor (2025-04-12):

![Screenshot of the block editor with two notices about a missing main element](https://github.com/user-attachments/assets/1306388a-3011-42a9-8347-5f117e44907e)


Old screenshots:

No warnings:
![The document outline without warnings, showing the main and a few headings](https://github.com/user-attachments/assets/83c1c27c-d93f-40c8-a989-3f5fe1fdb6da)

Neither main or headings:
<img width="351" alt="Document Outline warning for no main element and no headings" src="https://github.com/user-attachments/assets/44af1044-8d80-4029-9ee4-c09b40d9e3e5" />

A heading, but no main:
<img width="351" alt="Document Outline warning for no main element" src="https://github.com/user-attachments/assets/c20b174d-07d3-4267-af15-14b17a636167" />

Multiple main elements and incorrect heading order, with mains nested in mains:

<img width="351" alt="Document outline with multiple nested mains, with warnings showing" src="https://github.com/user-attachments/assets/157ec4a8-46d1-4321-bcf7-73c510769ab5" />

carolinan committed:
Document Outline: Add check for the main element

Mamaduka commented on packages/editor/src/components/document-outline/index.js:
```diff
@@ -154,9 +163,48 @@ export default function DocumentOutline( {
+	const classNames =
+		'document-outline' +
+		( headings.length < 1 ? ' has-no-headings' : '' ) +
+		( mainElements.length === 0 ? ' has-no-main' : '' );
```
You can use `clsx` library for this. Makes it easier to manage multiple classes.

carolinan replied:
Updated, thank you.

Mamaduka reviewed (commented):
I like this approach.

> I think this is an improvement but now that the main is showing, I would expect the outline to display the main element and headings in relation to each other.

This makes sense. The new outline will consist of headings and main element(s). I think this means that there will be a single `computeDocumentOutline` method that can handle both elements.

The `main` element warnings still only make sense for the `template` post type.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
Display nested blocks in the outline

carolinan committed:
Do not use the 'has-no-main' class name in the post editor

carolinan commented on packages/editor/src/components/document-outline/index.js:
```diff
@@ -113,105 +143,182 @@ export default function DocumentOutline( {
 					>
-						{ title }
+						{ incorrectMainTag }
 					</DocumentOutlineItem>
```
If I don't close this `</DocumentOutlineItem>` before listing the children, the `selectBlock` does not work on the children.
The result is that the children are not indented.

carolinan replied:
In other words this still require some work :)

carolinan committed:
Correct the isvalid value for the display of the Main element in the outline

Add a new constant named isValid that is true or false depending on if there is one main element.
Use isValid as a condition for displaying the warning text for the main element in the outline.

carolinan committed:
Multiple bug fixes

Combine the element count with the "headings by level" count.

Remove duplicate li elements

carolinan committed:
Indent nested items in the outline

carolinan commented:
> Question: Is updated headling nesting connected to the main elements? When multiple main element violations are fixed, the  document outline will no longer display that nesting, correct?

The visual change is only that the ul that wraps the children have some top and left margin.
When multiple main element violations are fixed, headings inside the main are still positioned with that margin, they show slightly to the left compared to headings that are outside the main.

The em dash that shows before the heading level is not part of this update, it looks like that on trunk as well.

carolinan commented:
Or maybe you are not thinking of the visual, but suggesting that I should flatten the list of blocks?

I tried it early on, and when I ran into issues I may - have gotten carried away and made the code too complex.

Mamaduka commented:
> Or maybe you are not thinking of the visual, but suggesting that I should flatten the list of blocks?

Exactly. Sorry if I wasn't clear.

carolinan committed:
Flatten the inner blocks and simplify the render

Also:
Add a TODO for the missing title, with a link to the open issue.
Remove unused CSS
Add parentheses around the warning message for the main, to match the headings.

carolinan commented:
I have flattened the inner blocks and simplified the counting of the elements and the render.

I kept two conditions that each add a `<DocumentOutlineItem>`, but this is for readability.
I personally found the code easier to read when the headings and mains were separate, because the code block for the heading warnings is so large [(lines)](https://github.com/WordPress/gutenberg/pull/68661/files#diff-bd465edcf520339c50446fa3c39bb014ec0f46cb9455075ccc0c775ffa9dbce0R266-R282).

carolinan commented:
It is assumed that the best practice is that the `<main>` element is added to the template.

But there is nothing preventing changing the HTML tag to a `<main>` on a group or cover that is placed in the post content.
So when the warnings related to the `<main>` does not show unless you are editing the template, there is a risk that duplicate mains are added.

Are there scenarios when the main needs to be added to the post content? Perhaps if the user does not have permission to edit templates?

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
List the main element if "show template" is enabled.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

luminuu commented:
I've checked out the pull request and it does work for me, however I'm a bit concerned it will not get any attention at all with the current styling. I've experimented a bit and have added the Notice component to the text, so it stands out much better if there's an issue:

![CleanShot 2025-02-05 at 18 24 06@2x](https://github.com/user-attachments/assets/43c67f8f-984c-4f96-9e07-9e1a252cf02a)

I don't think I can update the PR but I can share the code changes if we agree on this. @carolinan 

While this would be an improvement at this location, I wonder if the warning should be more prominent to the user without having to go into a certain part of the interface that is not visible by default. In the issue, I've added a screenshot of how it looks in the Greyd WP theme, here's an updated version with the current text and a link:

![CleanShot 2025-02-05 at 18 29 06@2x](https://github.com/user-attachments/assets/4c477075-e0f8-48d5-b25e-a4f3293b7886)

This message only shows if it's deteced on load and will be updated once the template it saved, the message will disappear if the problem was fixed.

carolinan commented:
Hi
The full width notice at the top still feels intrusive to me when I test it in the editor.
I wonder if the Document Outline tab label could have an icon (and screen reader alternative) that indicates if there are warnings.

afercia commented:
> I wonder if the Document Outline tab label could have an icon (and screen reader alternative) that indicates if there are warnings.

That's an interesting idea worth exploring. 💡  Today it's a common pattern to communicate a notification. I would say that when the document outline contains a warning that is buried down within the popover and Outline tab, users should be notified by the means of a red dot or something similar directly in the top bar. Of course, the information should be communicated in an accessible way as well.

![Screenshot 2025-02-06 at 08 28 55](https://github.com/user-attachments/assets/49593b78-d3df-478b-a1a7-b75d310721c0)

luminuu commented:
I like the idea of the notification indicator! Should we try to implement it with this PR or create a separate issue+PR for that?

@carolinan Do we want to stick with the Notice component with the message though? Like in my first screenshot of my last post. Otherwise with just light text it's not really recognisable as a notice.

carolinan commented:
I think it may need to be a separate issue

luminuu commented on packages/editor/src/components/document-outline/index.js:
```diff
@@ -113,104 +148,145 @@ export default function DocumentOutline( {
+				</>
+			) }
+			{ isShowingTemplate && mainElements.length === 0 && (
 				<p>
```
I'd suggest to wrap this warning in a `Notice` component, which is included in `@wordpress/components`. Here's the source code I've tested it with locally:

```
<Notice
spokenMessage={ null }
status="warning"
isDismissible={ false }
>
<p>
	{ __(
		'The main element is missing. Select the block that contains your most important content and add the main HTML element in the Advanced panel.'
	) }
</p>
</Notice>
```

luminuu commented on packages/editor/src/components/document-outline/style.scss:
```diff
@@ -82,7 +82,8 @@
 
-.editor-document-outline.has-no-headings {
+.document-outline.has-no-headings,
+.document-outline.has-no-main {
```
If the warning will be in the `Notice` component, this CSS change is not needed anymore.

luminuu reviewed (changes requested):
I have added my suggestion for the using the `Notice` component to move this PR forward.

carolinan commented:
Thank you. The reason why I have been holding this PR off is to reduce the conflicts with https://github.com/WordPress/gutenberg/pull/69073/

luminuu commented:
I've now read through the other issue + PR and I don't think these two depend too much on each other, as the main tag warning should only show in the site editor in my opinion. Because on block editor level, this would not make any sense, as the content shown in there would ideally always be inside the main tag.

I'm not sure if we need to show this warning if the "show template" option is activated. If this is the future goal to combine both editors it would make sense, but from a current point of view it makes most sense in the site editor only. Correct me if I'm wrong on this.

carolinan committed:
Try to solve merge conflicts

carolinan committed:
Use a Notice to display the warning about the missing main element.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan commented:
I have updated the screenshots of the PR to show what it looks like now.

@luminuu I have implemented the Notice for the message that shows when there is no `<main>.`
But what about when there are multiple?
(I think at least the parenthesis should be removed)

carolinan commented:
> I've now read through the other issue + PR and I don't think these two depend too much on each other, as the main tag warning should only show in the site editor in my opinion. Because on block editor level, this would not make any sense, as the content shown in there would ideally always be inside the main tag.
> 
> I'm not sure if we need to show this warning if the "show template" option is activated. If this is the future goal to combine both editors it would make sense, but from a current point of view it makes most sense in the site editor only. Correct me if I'm wrong on this.

Ideally it should be in the template. But there is nothing preventing the user from adding the `<main>` in the post content and if they do, there will be more than one, and it should warn against it.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

karmatosed commented:
One thing that strikes me about this is that a double placement approach could be very useful. What if it not only appears in the editor space with a dismissible notice, but also right by where you could change the HTML element? This builds on the pattern that accessibility has for contrast.

<img width="1607" alt="image" src="https://github.com/user-attachments/assets/b5bc140b-018d-4e4f-a71b-c9df701aa452" />

I would recommend this over putting it in the document outline.

carolinan commented:
I will try to make the time to fix the merge conflict and experiment with the advanced panel idea + re-add the full width notice in the editor this week.

carolinan commented:
Perhaps obvious, but the HTML Element control in Advanced panel is only a `<SelectControl>` component added to the individual blocks.
That makes it difficult to add any validation of the HTML tags in the control.
In my opinion, _it could be done by creating a new component_ and using that in these blocks, but not as part of this PR.

carolinan committed:
An attempt to reduce the merge conflicts

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
Continue solving merge conflicts

carolinan commented:
I am running into an issue where the notices for the heading level order does not show after headings are moved unless I close and re-open the document outline.

carolinan committed:
Add the notice about the missing main element at the top of the editor

carolinan commented:
I was trying to add a hook to the `editor` package so that the notice could show without the Document Outline open, but I was not able to make it work: The notice re-rendered too many times and was not dismissible. So that was a couple of hours wasted.

So now the notice at the top of the editor only shows once the Document Outline is opened by the user.
✅   If a main element exists but is removed, the notice appears.
❌   If a main element does not exist but is added, the notice does not disappear (unless dismissed).

carolinan committed:
Add a link to a HTML reference about the main element

carolinan commented on packages/editor/src/components/document-outline/index.js:
```diff
@@ -147,35 +201,72 @@ export default function DocumentOutline( {
+			createNotice(
+				'warning',
+				__(
+					'The main HTML element is missing. Select the block that contains your most important content and add the main HTML element in the Advanced panel.'
```
`createNotice ` can not include the link to the HTML reference, so we end up with two very similar text strings.

carolinan commented:
I have not been able to solve this:
❌ If a main element does not exist but is added, the notice does not disappear (unless dismissed).

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
Fix issue with notice not disappearing when main element is added

karmatosed commented:
I would like to note that my suggestion isn't really resolved here: https://github.com/WordPress/gutenberg/pull/68661#issuecomment-2791077640. I understand the issue but what I am suggesting is an improved user experience over the one in this PR. Which does need design feedback. 

Whilst I do respect and value this perspective:

> There are issues with adding elements to the Advanced panel conditionally because the order in the panel is not guaranteed. We learnt this while adding the aria-label to the group block. It was not possible to "group" the aria-label input field with the HTML element select list. To do that we would need to create a new Slot and display it for example below the Advanced panel.

I think we need to review approaches here and the experience before merging.

That said, as with anything there is always a middle ground and a solution that can balance both. So I wonder what could do that?

carolinan commented:
I do not agree with removing the check from the document outline.

carolinan commented:
I did not understand this sentence.
> If there are messages that go beyond main then absolutely they belong outside there if can't be then set,


If you are asking if there are other error messages In the document outline than the missing or duplicate main, then yes, there are. 
There are warnings for:
- Incorrect heading order
- Multiple H1 heading.

carolinan commented:
If you are saying that whenever I select a block, there is a message in the advanced panel that says that the main element is missing, then yes I think that is a bad idea.

1) Because as soon as the user sees that message, they are gong to set the `<main`> on whatever block they have currently selected.
The `<main>` is going to end up wrapping the header. Or the featured image. Or the navigation.
That is not accessible. The `<main>` needs to wrap the most important content.

2) The advanced panel also isn't discoverable.

carolinan commented:
How do we unblock this?
As I wrote I do not understand what it is you want to show here. All I see is an arrow.
Ad it means I also can not contribute to building it, neither here or as a separate GitHub issue.


> One thing that strikes me about this is that a double placement approach could be very useful. What if it not only appears in the editor space with a dismissible notice, but also right by where you could change the HTML element? This builds on the pattern that accessibility has for contrast.
> 
> <img alt="image" width="1607" src="https://private-user-images.githubusercontent.com/253067/432047375-b5bc140b-018d-4e4f-a71b-c9df701aa452.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMTI2MzAsIm5iZiI6MTc1MjExMjMzMCwicGF0aCI6Ii8yNTMwNjcvNDMyMDQ3Mzc1LWI1YmMxNDBiLTAxOGQtNGU0Zi1hNzFiLWM5ZGY3MDFhYTQ1Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMFQwMTUyMTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NjRiNjQzMDlmMjdkMTMyNzA0OTFjZWY4YmQzOWVjOWQwNTMzYzJhMjdlZWVkZjA2OGQwYTM4MTEzOGUzNmRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.i4z3LBFrFkas-X1hS3YKASN6Dbd6GSk7Z9ls6bFeoaM">
> I would recommend this over putting it in the document outline.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

Mamaduka committed:
Merge branch 'trunk' into add/document-outline-main

Mamaduka committed:
Fix lint error

Mamaduka commented:
@karmatosed, I believe what you've proposed in https://github.com/WordPress/gutenberg/pull/68661#issuecomment-2791077640 has already been implemented via #69904.

Mamaduka committed:
Merge branch 'trunk' into add/document-outline-main

Mamaduka committed:
Fix text alignimet for overview notice

Mamaduka committed:
Swap illustration and notice order

Mamaduka committed:
Remove line break for incorrectMainTag message

Mamaduka committed:
Remove global notice

karmatosed commented:
Thanks for clarity there @Mamaduka and @jasmussen I would appreciate a second opinion on this too, my biggest concern is we don't get into over information states.

> Maybe we can reduce visual noise by making the following changes:

> Keep the main tag list item text brief. Display a block name or something similar.
> Change the warning text based on type: a) the main tag is missing, b) multiple main tags in the document.

This feels better to me if everyone agrees.

jasmussen commented:
No blocking feedback, PR feels good and valid. I'd agree that it could be polished a bit in terms of verbiage and otherwise, but I don't know if it has to block the PR or happen separately, as the order change seems fine.

* I'm not sure why the parenthesis is necessary for the explanation of the missing element, may work without it. 
* I'm waffling on whether the instruction (Adjust the HTML element) is useful in the italic description or not. Perhaps it's okay to remove if the instruction remains, somehow, in the notice.

Almost certainly a separate concern and not necessarily something to fix here. Notice doesn't feel like the perfect way to frame the message. I think of this panel as a "linting" panel, which already now means there's a bit of a disconnect in that it looks different if it's "empty" vs. not empty. Perhaps we should embrace it more as a console, and when it's empty, it's just empty. And relatedly, if we do treat it as a linting console, it furthers that Notice may not be the right component. I think of this more as a _list of linting things to fix_, and that they should come with three levels of warning. Notice (`console.log( ... );`), warning, and error, with similarly three levels of emphasis.

Probably needs a design, and again, not blocking.

carolinan committed:
Merge branch 'trunk' into add/document-outline-main

carolinan committed:
Merge branch 'trunk' into add/document-outline-main
